crates/viewer/re_viewer_context/src/cache/video_stream_cache.rs

```diff
@@ -0,0 +1,555 @@
+    // TODO(andreas): Should validate whether all codecs ever logged are the same, but it's a bit tedious.
+    let last_codec = codec_chunks
+        .last()
+        .and_then(|chunk| {
+            chunk.component_instance::<components::VideoCodec>(&codec_chunk_descr, 0, 0)
+        })
+        .ok_or(VideoStreamProcessingError::MissingCodec)?
+        .map_err(VideoStreamProcessingError::FailedReadingCodec)?;
+    let codec = match last_codec {
+        components::VideoCodec::H264 => re_video::VideoCodec::H264,
+        // components::VideoCodec::H265 => re_video::VideoCodec::H265,
+        // components::VideoCodec::VP8 => re_video::VideoCodec::Vp8,
+        // components::VideoCodec::VP9 => re_video::VideoCodec::Vp9,
+        // components::VideoCodec::AV1 => re_video::VideoCodec::Av1,
+    };
+
+    // Extract all video samples.
+    let mut video_sample_buffers = StableIndexDeque::new();
+    let mut samples = StableIndexDeque::with_capacity(video_chunks.len()); // Number of video chunks is minimum number of samples.
+    let mut gops = StableIndexDeque::new();
+
+    for chunk in video_chunks {
+        read_samples_from_chunk(
+            timeline,
+            &frame_sample_descr,
+            chunk,
+            codec,
+            &mut video_sample_buffers,
+            &mut samples,
+            &mut gops,
+        )?;
+    }
+
+    Ok((
+        re_video::VideoDataDescription {
+            codec,
+            stsd: None,                               // Only on mp4s have this.
+            coded_dimensions: None,                   // Unknown so far.
+            timescale: re_video::Timescale::IDENTITY, // We use raw rerun timestamps, so we don't have to scale time.
+            duration: None, // Streams have to be assumed to be open ended, so we don't have a duration.
+            gops,
+            samples,
+            samples_statistics: re_video::SamplesStatistics::NO_BFRAMES, // TODO(#10090): No b-frames for now.
+            mp4_tracks: std::iter::once((0, Some(re_video::TrackKind::Video))).collect(),
+        },
+        video_sample_buffers,
+    ))
+}
+
+/// Reads all samples from a chunk and appends them to a list.
+///
+/// Rejects out of order samples - new samples must have a higher timestamp than the previous ones.
+fn read_samples_from_chunk(
+    timeline: TimelineName,
+    frame_chunk_descr: &re_types::ComponentDescriptor,
+    chunk: &re_chunk::Chunk,
+    codec: re_video::VideoCodec,
+    chunk_buffers: &mut StableIndexDeque<SampleBuffer>,
+    samples: &mut StableIndexDeque<re_video::SampleMetadata>,
+    gops: &mut StableIndexDeque<re_video::demux::GroupOfPictures>,
+) -> Result<(), VideoStreamProcessingError> {
+    re_tracing::profile_function!();
+
+    let Some(raw_array) = chunk.raw_component_array(frame_chunk_descr) else {
+        // This chunk doesn't have any video chunks.
+        return Ok(());
+    };
+
+    let inner_list_array = raw_array
```

Comment:
This function badly need a long explanation and/or a diagram. Here's a start:

The `raw_array` above is logically a `Vec<Vec<Blob>>`, where the inner `Vec` always has a len=1, because we're dealing with a "mono-component" (each `VideoStream` has exactly one `VideoSample` instance per time)`.

Because of how arrow works, the bytes of all the blobs are actually sequential in memory (yay!) in a single buffer, what you call `values` below (could use a better name btw).

We want to figure out the byte offsets of each blob within the arrow buffer that holds all the blobs, i.e. get out a `Vec<ByteRange>`.